### DISCUSSION TITLE
Differentiable Colour Difference Functions

### DISCUSSION THREAD

**Discussion by JoniSuominen** [colour-science/colour#1254] [Ideas]
Hi!

I recently re-implemented a neural network from this paper: https://doi.org/10.2352/issn.2694-118X.2021.LIM-54 in Python (specifically PyTorch) for my Master's thesis. Unfortunately, I could not use the colour difference functions (e.g. CIEDE2000) implemented in this library, as they operate on Numpy arrays rather than PyTorch tensors.

I had an idea to make an ML library for colour science functions that implement basic conversions and colour difference functions for the most popular frameworks (TensorFlow and PyTorch). At the moment there is no "open-source" library to my knowledge for this, although ML/DL is a growing field. There are individual packages, such as: https://github.com/jorge-pessoa/pytorch-colors and https://github.com/ZhengyuZhao/PerC-Adversarial/tree/master, but I figured it would be worth the effort.

Would this be something you support, maybe as another library in the colour-science family?

**Comment by thomasmansencal** [comment]
Hello @JoniSuominen,

This is a timely discussion! Beyond PyTorch tensors, we are looking at supporting the Python Array API Standard as per #1244.
Work has not started yet but this is some of the highest priority issue at this point and we will definitely get there. Feel free to chime in the #1244 thread.

Cheers,

Thomas

**Reply by JoniSuominen** [reply]
Thanks! I'll take a look and would be very happy to contribute!

Best,
Joni

**Comment by edwinksl** [comment]
Similar to @JoniSuominen, I was considering using the `delta_E_CIE2000` function directly as a loss function in my PyTorch model. Hoping to see such a capability being implemented soon!
